src/ledger/DataFrame.cpp

```diff
@@ -0,0 +1,293 @@
+    auto& st = prep.statement();
+    st.exchange(use(actIDStrKey));
+    st.exchange(use(dataName));
+
+    auto timer = db.getSelectTimer("data");
+    loadData(prep, [&retData](LedgerEntry const& data)
+               {
+                   retData = make_shared<DataFrame>(data);
+               });
+
+    return retData;
+}
+
+void
+DataFrame::loadData(StatementContext& prep,
+                       std::function<void(LedgerEntry const&)> dataProcessor)
+{
+    string actIDStrKey;
+   
+    std::string dataName,dataValue;
+
+    soci::indicator dataNameIndicator, dataValueIndicator;
+
+    LedgerEntry le;
+    le.data.type(DATA);
+    DataEntry& oe = le.data.data();
+
+    statement& st = prep.statement();
+    st.exchange(into(actIDStrKey));
+    st.exchange(into(dataName, dataNameIndicator));
+    st.exchange(into(dataValue, dataValueIndicator));
+    st.define_and_bind();
+    st.execute(true);
+    while (st.got_data())
+    {
+        oe.accountID = PubKeyUtils::fromStrKey(actIDStrKey);
+        
+        if((dataNameIndicator != soci::i_ok) ||
+            (dataValueIndicator != soci::i_ok))
+        {
+            throw std::runtime_error("bad database state");
+        }
+        oe.dataName = dataName;
+        bn::decode_b64(dataValue, oe.dataValue);
+
+        dataProcessor(le);
+        st.fetch();
+    }
+}
+
+
+
+void
+DataFrame::loadAccountsData(AccountID const& accountID,
+                       std::vector<DataFrame::pointer>& retData,
+                       Database& db)
+{
+    std::string actIDStrKey;
+    actIDStrKey = PubKeyUtils::toStrKey(accountID);
+
+    std::string sql = dataColumnSelector;
+    sql += " WHERE accountid = :id";
+    auto prep = db.getPreparedStatement(sql);
+    auto& st = prep.statement();
+    st.exchange(use(actIDStrKey));
+
+    auto timer = db.getSelectTimer("data");
+    loadData(prep, [&retData](LedgerEntry const& of)
+               {
+                   retData.emplace_back(make_shared<DataFrame>(of));
+               });
+}
+
+std::unordered_map<AccountID, std::vector<DataFrame::pointer>>
+DataFrame::loadAllData(Database& db)
+{
+    std::unordered_map<AccountID, std::vector<DataFrame::pointer>> retData;
+    std::string sql = dataColumnSelector;
+    sql += " ORDER BY accountid";
+    auto prep = db.getPreparedStatement(sql);
+
+    auto timer = db.getSelectTimer("data");
+    loadData(prep, [&retData](LedgerEntry const& of)
+               {
+                   auto& thisUserData = retData[of.data.data().accountID];
+                   thisUserData.emplace_back(make_shared<DataFrame>(of));
+               });
+    return retData;
+}
+
+bool
+DataFrame::exists(Database& db, LedgerKey const& key)
+{
+    std::string actIDStrKey = PubKeyUtils::toStrKey(key.data().accountID);
+    std::string dataName = key.data().dataName;
+    int exists = 0;
+    auto timer = db.getSelectTimer("data-exists");
+    auto prep =
+        db.getPreparedStatement("SELECT EXISTS (SELECT NULL FROM accountdata "
```

Comment:
What does the execution plan look like on this?  Any reason not to simply use `SELECT 1 from accountdata WHERE account=:id AND dataname=:s`?

I'm always wary of subqueries, especially when aiming against multiple database engines.